File: baekjoon/C++/1486.cc
#include <iostream>
#include <vector>
#include <algorithm>
#include <queue>
#include <string>
#include <cmath>  
#define endl '\n'
using namespace std;
using pii = pair<int, int>;

int x_move[] = {-1, 1, 0, 0};
int y_move[] = {0, 0, -1, 1};
int Row, Col, T, D; 
vector<vector<int>> matrix;

void input()
{
    cin >> Row >> Col >> T >> D;
    string str;

    matrix.assign(Row + 1, vector<int>(Col + 1, 0));
    for (int i = 1; i <= Row; i++)
    {
        cin >> str;
        for (int j = 0; j < str.size(); j++)
        {
            if (str[j] >= 'a' && str[j] <= 'z')
                matrix[i][j + 1] = str[j] - 'a' + 26;
            else
                matrix[i][j + 1] = str[j] - 'A';
        }
    }
}

bool inTheMatrix(int xpos, int ypos)
{
    return xpos >= 1 && xpos <= Row && ypos >= 1 && ypos <= Col;
}

vector<vector<int>> Dijkstra(pair<int, int> start)
{
    vector<vector<int>> dis(Row + 1, vector<int>(Col + 1, 1e9));
    dis[start.first][start.second] = 0;
    priority_queue<pair<int, pii>, vector<pair<int, pii>>, greater<pair<int, pii>>> pq; // {cost, {x, y}}

    pq.push({0, start});

    while (!pq.empty())
    {
        int cost = pq.top().first;
        int xpos = pq.top().second.first;
        int ypos = pq.top().second.second;
        pq.pop();

        if (dis[xpos][ypos] < cost)
            continue;

        for (int i = 0; i < 4; i++)
        {
            int x = xpos + x_move[i];
            int y = ypos + y_move[i];

            if (!inTheMatrix(x, y))
                continue;

            int height_diff = matrix[x][y] - matrix[xpos][ypos];

            if (abs(height_diff) > T)
                continue;

            int weight = (height_diff > 0) ? pow(height_diff, 2) : 1;
            int new_cost = dis[xpos][ypos] + weight;

            if (dis[x][y] > new_cost)
            {
                dis[x][y] = new_cost;
                pq.push({new_cost, {x, y}});
            }
        }
    }
    return dis;
}

void solve()
{
    input();

    vector<vector<int>> go_dis = Dijkstra({1, 1});

    int max_height = matrix[1][1]; 

    for (int i = 1; i <= Row; i++)
    {
        for (int j = 1; j <= Col; j++)
        {
            if (go_dis[i][j] == 1e9 || go_dis[i][j] > D)
                continue; 

            vector<vector<int>> return_dis = Dijkstra({i, j});
            if (return_dis[1][1] != 1e9 && go_dis[i][j] + return_dis[1][1] <= D)
            {
                max_height = max(max_height, matrix[i][j]);
            }
        }
    }

    cout << max_height << endl;
}

int main(void)
{
    ios::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);

    solve();
    return (0);
}

File: baekjoon/java/1486.java
/**
 * A ~ Z (0,25), a ~ z (26,51)
 * 차이 x <= T
 *
 * 낮은 곳이나 같은 곳 -> 시간은 1초
 * 높은 곳으로 가면 차이의 제곱
 *
 * 어두워지는 시간 D
 * 올라갈 수 있는 최대의 높이
 *
 * 갔다가 돌아오는 것은 시간이 다르다
 *
 * 1 2 3
 * 3 4 5
 *
 * 0 1 2
 * 4 5 6
 *
 *
 *
 *
 */

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class Main {

    static class Edge implements Comparable<Edge>{
        int xpos;
        int ypos;
        long dis;


        Edge (int xpos, int ypos, long dis) {
            this.xpos = xpos;
            this.ypos = ypos;
            this.dis = dis;
        }

        @Override
        public int compareTo(Edge o) {
            return Long.compare(this.dis, o.dis);
        }
    }


    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static int Row, Col, T, D;
    static int[][] matrix;
    static final int[] x_move = {-1,1,0,0}, y_move = {0,0,-1,1};
    static long[][] dis1, dis2;
    static int res;

    public static void main(String[] args) throws Exception{
        input();
        Dijkstra(1,1, dis1, false);
        Dijkstra(1, 1, dis2, true);


        //showMatrix(dis1);
        //System.out.println();
        //showMatrix(dis2);

        solve();

        System.out.println(res);
    }
    static void solve() {
        for (int i = 1; i <= Row; ++i) {
            for (int j = 1; j <= Col; ++j) {
                if (dis1[i][j] == Long.MAX_VALUE) continue;
                if (dis2[i][j] == Long.MAX_VALUE) continue;

                if (dis1[i][j] + dis2[i][j] <= D) {
                    res = Math.max(res, matrix[i][j]);
                }
            }
        }
    }
    static void showMatrix() {
        for (int i = 1; i <= Row; ++i) {
            System.out.println(Arrays.toString(matrix[i]));
        }
    }
    static void showMatrix(long[][] matrix) {
        for (int i = 1; i <= Row; ++i) {
            System.out.println(Arrays.toString(matrix[i]));
        }
    }



    static void input() throws Exception {
        StringTokenizer st = new StringTokenizer(br.readLine());
        Row = Integer.parseInt(st.nextToken());
        Col = Integer.parseInt(st.nextToken());
        T = Integer.parseInt(st.nextToken());
        D = Integer.parseInt(st.nextToken());

        matrix = new int[Row+1][Col+1];
        dis1 = new long[Row+1][Col+1];
        dis2 = new long[Row+1][Col+1];

        for (int i = 1; i <= Row; ++i) {
            String str = br.readLine();
            for (int j = 1; j <= Col; ++j) {
                char c = str.charAt(j-1);

                if (Character.isUpperCase(c)) {
                    matrix[i][j] = (int) c - 'A';
                } else {
                    matrix[i][j] = (int) c - 'a' + 26;
                }

            }
        }
    }

    static boolean inTheMatrix(int xpos, int ypos) {
        return xpos >= 1 && xpos <= Row && ypos >= 1 && ypos <= Col;
    }


    static void Dijkstra(int sx, int sy, long[][] dis, boolean reverse) {
        for (int i = 1; i <= Row; ++i) Arrays.fill(dis[i], Long.MAX_VALUE);

        PriorityQueue<Edge> pq = new PriorityQueue<>();
        pq.add(new Edge(sx, sy, 0));
        dis[sx][sy] = 0;

        while (!pq.isEmpty()) {
            Edge e = pq.poll();
            int x = e.xpos, y = e.ypos;
            long d = e.dis;

            if (d != dis[x][y]) continue;

            for (int k = 0; k < 4; ++k) {
                int nx = x + x_move[k];
                int ny = y + y_move[k];

                if (!inTheMatrix(nx, ny)) continue;
                if (Math.abs(matrix[nx][ny] - matrix[x][y]) > T) continue;

                long w;
                if (!reverse) {
                    if (matrix[nx][ny] <= matrix[x][y]) w = 1;
                    else {
                        long diff = matrix[nx][ny] - matrix[x][y];
                        w = diff * diff;
                    }
                } else {
                    if (matrix[x][y] <= matrix[nx][ny]) w = 1;
                    else {
                        long diff = matrix[x][y] - matrix[nx][ny];
                        w = diff * diff;
                    }
                }

                long nd = dis[x][y] + w;
                if (nd < dis[nx][ny]) {
                    dis[nx][ny] = nd;
                    pq.add(new Edge(nx, ny, nd));
                }
            }
        }
    }


}
